### PR TITLE
Allow custom HTML in HTML Emails

### PR DESCRIPTION
**Pull request by nickvergessen**
We need to allow custom HTML, so e.g. the activity app can link file names to the file in the web view.
Currently that is not possible because all HTML is escaped.

The question is if we want to do it the simple way like this, or if we want to add a flag/new method that does not automatically-escape the input.

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Then we need to check if all callers properly escape this and document this in the PHPDoc of the method, but I'm fine with that to not clutter the API even more.

**Comment by rullzer** [comment]
Mmmm yes. lets do it like this for now we can always improve later.

**Comment by nickvergessen** [comment]
@MorrisJobke can you generate a list for the appstore submissions?

**Comment by nickvergessen** [comment]
which apps are using that method and might need adjustment

**Comment by MorrisJobke** [comment]
`addBodyText`:

bildschirmfoto 2018-02-01 um 14 17 24

`addBodyListItem`:

bildschirmfoto 2018-02-01 um 14 19 34

`addBodyButton`:

bildschirmfoto 2018-02-01 um 14 20 52

**Comment by nickvergessen** [comment]
Hmm while looking into a patch for the calendar, I realize why a parameter or second method would be better:

```php
// before
$emailTemplate->addBodyText($this->l10n->t('We wanted to inform you that %s has published the calendar »%s«.', [$displayName, $calendarName]));

// after
$emailTemplate->addBodyText(
    htmlspecialchars($this->l10n->t('We wanted to inform you that %s has published the calendar »%s«.', [$displayName, $calendarName])),
    $this->l10n->t('We wanted to inform you that %s has published the calendar »%s«.', [$displayName, $calendarName])
);
```

basically we need to touch all usages of those methods which use fallback `$plain = ''`.

Or we automatically encode when plain is the fallback, because that means you want to display the same string and there should not be HTML inside, so we need to special char it.

**Comment by MorrisJobke** [comment]
> Hmm while looking into a patch for the calendar, I realize why a parameter or second method would be better:

I'm fine with that. You need to provide the HTML and the plain text separately. If you want to put HTML in there it should also not present in the plain text (or at least in a different way). So this is okay. Let's go for the "two strings needs to be passed" method I would say.

**Review by MorrisJobke — APPROVED** [review]

**Comment by juliushaertl** [comment]
> Or we automatically encode when plain is the fallback, because that means you want to display the same string and there should not be HTML inside, so we need to special char it.

I think a fallback makes sense here - use the text for both and escape the html if only one string is provided.

File: lib/private/Mail/EMailTemplate.php [hunk 1]
```diff
@@ -420,7 +420,7 @@ protected function ensureBodyIsOpened() {
 	/**
 	 * Adds a paragraph to the body of the email
 	 *
-	 * @param string $text
+	 * @param string $text Note: When $plainText falls back to this, HTML is automatically escaped in the HTML email
```
Comment (MorrisJobke) [inline lib/private/Mail/EMailTemplate.php]: Please put this also in the public interface ;)

**Review by MorrisJobke — APPROVED** [review]

**Review by rullzer — APPROVED** [review]

**Pull request merged**